test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/PriorityQueueTest.scala

```diff
@@ -236,4 +241,192 @@ class PriorityQueueTest {
     assertEquals(-0.987, pqDouble.poll(), 0.0)
   }
 
+  @Test def poll(): Unit = {
+    val pq = newPriorityQueueWith0Until100()
+
+    var nextExpected = 0
+    while (!pq.isEmpty()) {
+      assertEquals(nextExpected, pq.poll())
+      nextExpected += 1
+    }
+  }
+
+  @Test def removingAnArbitraryElementPreservesPriorities(): Unit = {
+    for (itemToRemove <- 0 until 100) {
+      val pq = newPriorityQueueWith0Until100()
+
+      assertTrue(pq.remove(itemToRemove))
+
+      var nextExpected = 0
+      while (!pq.isEmpty()) {
+        if (nextExpected == itemToRemove)
+          nextExpected += 1
+        assertEquals(s"after removing $itemToRemove", nextExpected, pq.poll())
+        nextExpected += 1
+      }
+    }
+  }
+
+  @Test def removingAnArbitraryElementWithAnIteratorPreservesPriorities(): Unit = {
+    for (itemToRemove <- 0 until 100) {
+      val pq = newPriorityQueueWith0Until100()
+
+      val iter = pq.iterator()
+      while (iter.next() != itemToRemove) {}
+      iter.remove()
+
+      var nextExpected = 0
+      while (!pq.isEmpty()) {
+        if (nextExpected == itemToRemove)
+          nextExpected += 1
+        assertEquals(s"after removing $itemToRemove", nextExpected, pq.poll())
+        nextExpected += 1
+      }
+    }
+  }
+
+  @Test def removingAnArbitraryElementWithAnIteratorPreservesPrioritiesCornerCase(): Unit = {
+    /* This test depends on internal details of our implementation. It may have
+     * to be changed in the internal data structure and/or algorithms of
+     * PriorityQueue are modified in the future.
+     */
```

Comment:
I'm not sure I understand why these tests must be whitebox tests in order to test what they test. Could you clarify this in the comments? (or if they don't have to, change them to blackbox tests)?

Reply:
They don't need to be whitebox to test what they test. However, they need to be whitebox to *make sure* that they test what we think they're testing. In particular, the scenario where `Iterator.remove()` causes reordering in the heap such that elements yet to be iterated are moved before the cursor, and elements already iterated are moved after. This scenario is not easy to trigger. It requires some deliberate engineering of how the heap will look like and what specific element will be removed.